torchani/cuaev/aev.h

```diff
@@ -0,0 +1,229 @@
+//   \    /                \                   ^
+//  [dddist] $$$$ $$$$ [ddangular] $$$$  $$$$  ^
+//       \   /    /           \      |    /    ^
+//        \_/____/             \_____|___/     ^
+//         |    _____________________/         ^
+//         |   /                               ^
+//      [ddcoord]                              ^
+//         |                                   ^
+//        ...                                  ^
+//         |                                   ^
+//       [dout2]                               ^
+//
+// Functional relationship:
+// dout2 <-- input
+// ddcoord(dout2, ...)
+// dddist = dist_doublebackward(ddcoord, coord, dist)
+// ddradial = radial_doublebackward(dddist, dist)
+// ddangular = angular_doublebackward(ddcord, coord, dist)
+// ddaev = concatenate(ddradial, ddangular)
+// dparams(ddaev, ...) <-- output
+
+struct AEVScalarParams {
+  float Rcr;
+  float Rca;
+  int radial_sublength;
+  int radial_length;
+  int angular_sublength;
+  int angular_length;
+  int num_species;
+  Tensor EtaR_t;
+  Tensor ShfR_t;
+  Tensor EtaA_t;
+  Tensor Zeta_t;
+  Tensor ShfA_t;
+  Tensor ShfZ_t;
+  AEVScalarParams(
+      float Rcr_,
+      float Rca_,
+      Tensor EtaR_t_,
+      Tensor ShfR_t_,
+      Tensor EtaA_t_,
+      Tensor Zeta_t_,
+      Tensor ShfA_t_,
+      Tensor ShfZ_t_,
+      int num_species_);
+};
+
+struct Result : torch::CustomClassHolder {
+  Tensor aev_t;
+  Tensor tensor_Rij;
+  Tensor tensor_radialRij;
+  Tensor tensor_angularRij;
+  int total_natom_pairs;
+  int nRadialRij;
+  int nAngularRij;
+  Tensor tensor_centralAtom;
+  Tensor tensor_numPairsPerCenterAtom;
+  Tensor tensor_centerAtomStartIdx;
+  int maxnbrs_per_atom_aligned;
+  int angular_length_aligned;
+  int ncenter_atoms;
+  Tensor coordinates_t;
+  Tensor species_t;
+
+  Result(
+      Tensor aev_t_,
+      Tensor tensor_Rij_,
+      Tensor tensor_radialRij_,
+      Tensor tensor_angularRij_,
+      int64_t total_natom_pairs_,
+      int64_t nRadialRij_,
+      int64_t nAngularRij_,
+      Tensor tensor_centralAtom_,
+      Tensor tensor_numPairsPerCenterAtom_,
+      Tensor tensor_centerAtomStartIdx_,
+      int64_t maxnbrs_per_atom_aligned_,
+      int64_t angular_length_aligned_,
+      int64_t ncenter_atoms_,
+      Tensor coordinates_t_,
+      Tensor species_t_);
+  Result(tensor_list tensors);
+  tensor_list to_list();
+  void release();
+  ~Result() {
+    this->release();
+  }
+};
+
+// cuda kernels
+Result cuaev_forward(const Tensor& coordinates_t, const Tensor& species_t, const AEVScalarParams& aev_params);
+Tensor cuaev_backward(const Tensor& grad_output, const AEVScalarParams& aev_params, const Result& result);
+Tensor cuaev_double_backward(const Tensor& grad_force, const AEVScalarParams& aev_params, const Result& result);
+
+// CuaevComputer
+// Only keep one copy of aev parameters and one copy of result for backward
+struct CuaevComputer : torch::CustomClassHolder {
+  AEVScalarParams aev_params;
```

Comment:
If `AEVScalarParams` is the only data field of `CuaevComputer`, should we merge these classes?

Reply:
better not, aev_params is needed to pass to cuda_forward, backward, double_backward functions.

Reply:
Could we just pass CuaevComputer to cuda_forward, backward, and double_backward?